### ISSUE TITLE
Use an optimized alloc library

### ISSUE THREAD

**Issue by a-teammate** [inexor-game/code#159]
This could improve the speed of the overall system (especially loading):

http://locklessinc.com/benchmarks_allocator.shtml
best options are imo:
* http://gperftools.googlecode.com/svn/trunk/doc/tcmalloc.html
* http://www.canonware.com/jemalloc/
(both bsd licensed)

* ~~lockless~~ (GPL licensed)

Usage is simple: just link against it and it will override the default new and delete operators

**Comment by koraa** [comment]
Lockless seems to be the fastest allocator out there, as indicated by their own chart :)
Even though it's GPL3. it should still be OK to use. Zlib is in compatible with GPL and if we make the use optional there should be no problem removing this for closed source projects.

Did you check whether any of those provides c++ support? I think they only provide malloc/….
I suppose the default new operator and the C++ Allocator classes use malloc by default, but we would definitely need to check how exactly this behaves with the c++ constructs, our various libs and Windows.

**Comment by a-teammate** [comment]
All of those listed work natively with c++ code (just link against them) :) and at least tcmalloc is working under windows as well as far as i checked

**Comment by koraa** [comment]
how are they providing c++  support?

**Comment by a-teammate** [comment]
new and delete are wrappers for free and malloc, so theres no problem :) 
( e.g. last passage here explicitly lists effects for c++ http://locklessinc.com/install_windows.shtml )

However: the previous posted allocator benchmark is using outdated versions of those, meaning e.g. TCMalloc's most recent vesion from Jan 2015 or Jemalloc from april 2014 or even nedmalloc ( https://github.com/ned14/nedmalloc/ ) lastly updated july 2014 could outperform the quite outdated lockless from 2012 in the maintime.

**Comment by koraa** [comment]
And another one https://code.google.com/p/ltalloc/

**Comment by koraa** [comment]
I've just done some testing:

I tested the italloc and the GPL lockless allocator by loading them with LD_PRELOAD=. I found no perceived improvements.

In addition to that I also ran the google profiler and found that while loading maps standard malloc was used 0.2% of the time. Loading dust6 takes about 3s on my machine.
This means that malloc was active for about 6ms while loading data.

This is insubstantial, but I am not closing this ticket, because we are working on some other optimization and I am pretty sure that an improved malloc algorithm will matter in the future!

(https://code.google.com/p/gperftools/wiki/GooglePerformanceTools)
